File: cli/internal/connectapi/permissions.go
package connectapi

import (
	"context"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	appv1 "hmans.de/chatto/internal/pb/chatto/app/v1"
)

type permissionService struct {
	api *API
}

func (s *permissionService) GetRolePermissionTierMatrix(ctx context.Context, req *connect.Request[appv1.GetRolePermissionTierMatrixRequest]) (*connect.Response[appv1.GetRolePermissionTierMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetRolePermissionTierMatrix(ctx, caller.UserID, req.Msg.GetRoomId(), req.Msg.GetGroupId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.GetRolePermissionTierMatrixResponse{Matrix: apiTierRoles(matrix)}), nil
}

func (s *permissionService) GetRolePermissionMatrix(ctx context.Context, req *connect.Request[appv1.GetRolePermissionMatrixRequest]) (*connect.Response[appv1.GetRolePermissionMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetRolePermissionMatrix(ctx, caller.UserID, req.Msg.GetRoleName())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.GetRolePermissionMatrixResponse{Matrix: apiRolePermissionMatrix(matrix)}), nil
}

func (s *permissionService) GetUserPermissionMatrix(ctx context.Context, req *connect.Request[appv1.GetUserPermissionMatrixRequest]) (*connect.Response[appv1.GetUserPermissionMatrixResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	matrix, err := s.api.core.GetUserPermissionMatrix(ctx, caller.UserID, req.Msg.GetUserId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.GetUserPermissionMatrixResponse{Matrix: apiUserPermissionMatrix(matrix)}), nil
}

func (s *permissionService) ExplainPermissions(ctx context.Context, req *connect.Request[appv1.ExplainPermissionsRequest]) (*connect.Response[appv1.ExplainPermissionsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	explanations, err := s.api.core.ExplainPermissions(ctx, caller.UserID, req.Msg.GetUserId(), req.Msg.GetRoomId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.ExplainPermissionsResponse{Explanations: apiPermissionExplanations(explanations)}), nil
}

func (s *permissionService) SetRolePermission(ctx context.Context, req *connect.Request[appv1.SetRolePermissionRequest]) (*connect.Response[appv1.SetRolePermissionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	state, err := corePermissionState(req.Msg.GetDecision())
	if err != nil {
		return nil, err
	}
	scope := corePermissionTargetScope(req.Msg.GetScope())
	if err := s.api.core.SetRolePermissionState(ctx, caller.UserID, req.Msg.GetRoleName(), scope, core.Permission(req.Msg.GetPermission()), state); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.SetRolePermissionResponse{Ok: true}), nil
}

func (s *permissionService) RevokeRolePermissionGrant(ctx context.Context, req *connect.Request[appv1.RevokeRolePermissionGrantRequest]) (*connect.Response[appv1.RevokeRolePermissionGrantResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if err := s.api.core.AdminRevokeRolePermissionGrant(ctx, caller.UserID, req.Msg.GetRoleName(), core.Permission(req.Msg.GetPermission())); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.RevokeRolePermissionGrantResponse{Ok: true}), nil
}

func (s *permissionService) SetUserPermission(ctx context.Context, req *connect.Request[appv1.SetUserPermissionRequest]) (*connect.Response[appv1.SetUserPermissionResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	state, err := corePermissionState(req.Msg.GetDecision())
	if err != nil {
		return nil, err
	}
	scope := corePermissionTargetScope(req.Msg.GetScope())
	if err := s.api.core.SetUserPermissionState(ctx, caller.UserID, req.Msg.GetUserId(), scope, core.Permission(req.Msg.GetPermission()), state); err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&appv1.SetUserPermissionResponse{Ok: true}), nil
}

func apiPermissionExplanations(explanations []core.PermissionExplanation) []*appv1.PermissionExplanation {
	out := make([]*appv1.PermissionExplanation, 0, len(explanations))
	for i := range explanations {
		out = append(out, apiPermissionExplanation(explanations[i]))
	}
	return out
}

func apiPermissionExplanation(explanation core.PermissionExplanation) *appv1.PermissionExplanation {
	out := &appv1.PermissionExplanation{
		Permission:    string(explanation.Permission),
		State:         apiPermissionExplanationDecision(explanation.State),
		DecidedAt:     apiPermissionDecisionLevel(explanation.DecidedAt),
		DecidedByRole: explanation.DecidedByRole,
		Trace:         make([]*appv1.PermissionTraceEntry, 0, len(explanation.Trace)),
	}
	for i, entry := range explanation.Trace {
		out.Trace = append(out.Trace, &appv1.PermissionTraceEntry{
			Level:    apiPermissionDecisionLevel(entry.Level),
			RoleName: entry.RoleName,
			Decision: apiPermissionExplanationDecision(entry.Decision),
			Applied:  i == 0,
		})
	}
	return out
}

func apiPermissionExplanationDecision(decision core.DecisionKind) appv1.PermissionDecision {
	switch decision {
	case core.DecisionAllow:
		return appv1.PermissionDecision_PERMISSION_DECISION_ALLOW
	case core.DecisionDeny:
		return appv1.PermissionDecision_PERMISSION_DECISION_DENY
	default:
		return appv1.PermissionDecision_PERMISSION_DECISION_NONE
	}
}

func apiPermissionDecisionLevel(level core.PermissionLevel) appv1.PermissionDecisionLevel {
	switch level {
	case core.LevelServer:
		return appv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_SERVER
	case core.LevelGroup:
		return appv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_GROUP
	case core.LevelRoom:
		return appv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_ROOM
	default:
		return appv1.PermissionDecisionLevel_PERMISSION_DECISION_LEVEL_UNSPECIFIED
	}
}

func apiTierRoles(matrix *core.TierRoles) *appv1.TierRoles {
	if matrix == nil {
		return nil
	}
	out := &appv1.TierRoles{
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Roles:                 make([]*appv1.TierRole, 0, len(matrix.Roles)),
	}
	for _, role := range matrix.Roles {
		out.Roles = append(out.Roles, &appv1.TierRole{
			RoleName:         role.RoleName,
			DisplayName:      role.DisplayName,
			Description:      role.Description,
			IsSystem:         role.IsSystem,
			Position:         role.Position,
			Override:         apiTierPermissions(role.Override),
			InheritedAllows:  append([]string(nil), role.InheritedAllows...),
			InheritedDenials: append([]string(nil), role.InheritedDenials...),
		})
	}
	return out
}

func apiTierPermissions(perms core.TierPermissions) *appv1.TierPermissions {
	return &appv1.TierPermissions{
		Permissions:       append([]string(nil), perms.Permissions...),
		PermissionDenials: append([]string(nil), perms.PermissionDenials...),
	}
}

func apiRolePermissionMatrix(matrix *core.RolePermissionMatrix) *appv1.RolePermissionMatrix {
	if matrix == nil {
		return nil
	}
	return &appv1.RolePermissionMatrix{
		RoleName:              matrix.RoleName,
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Scopes:                apiPermissionMatrixScopes(matrix.Scopes),
		Cells:                 apiPermissionMatrixCells(matrix.Cells),
	}
}

func apiUserPermissionMatrix(matrix *core.UserPermissionMatrix) *appv1.UserPermissionMatrix {
	if matrix == nil {
		return nil
	}
	return &appv1.UserPermissionMatrix{
		UserId:                matrix.UserID,
		ApplicablePermissions: append([]string(nil), matrix.ApplicablePermissions...),
		Scopes:                apiPermissionMatrixScopes(matrix.Scopes),
		Cells:                 apiPermissionMatrixCells(matrix.Cells),
	}
}

func apiPermissionMatrixScopes(scopes []core.PermissionMatrixScope) []*appv1.PermissionMatrixScope {
	out := make([]*appv1.PermissionMatrixScope, 0, len(scopes))
	for _, scope := range scopes {
		out = append(out, &appv1.PermissionMatrixScope{
			Id:            scope.ID,
			Label:         scope.Label,
			Kind:          apiPermissionScopeKind(scope.Kind),
			ParentGroupId: scope.ParentGroupID,
		})
	}
	return out
}

func apiPermissionMatrixCells(cells []core.PermissionMatrixCell) []*appv1.PermissionMatrixCell {
	out := make([]*appv1.PermissionMatrixCell, 0, len(cells))
	for _, cell := range cells {
		out = append(out, &appv1.PermissionMatrixCell{
			Permission: cell.Permission,
			ScopeId:    cell.ScopeID,
			Override:   apiPermissionDecision(cell.Override),
			Effective:  apiPermissionDecision(cell.Effective),
		})
	}
	return out
}

func apiPermissionScopeKind(kind core.MatrixScopeKind) appv1.PermissionScopeKind {
	switch kind {
	case core.MatrixScopeGroup:
		return appv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_GROUP
	case core.MatrixScopeRoom:
		return appv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_ROOM
	default:
		return appv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_SERVER
	}
}

func apiPermissionDecision(decision core.MatrixDecision) appv1.PermissionDecision {
	switch decision {
	case core.MatrixDecisionAllow:
		return appv1.PermissionDecision_PERMISSION_DECISION_ALLOW
	case core.MatrixDecisionDeny:
		return appv1.PermissionDecision_PERMISSION_DECISION_DENY
	default:
		return appv1.PermissionDecision_PERMISSION_DECISION_NONE
	}
}

func corePermissionState(decision appv1.PermissionDecision) (core.PermissionState, error) {
	switch decision {
	case appv1.PermissionDecision_PERMISSION_DECISION_ALLOW:
		return core.PermissionStateAllow, nil
	case appv1.PermissionDecision_PERMISSION_DECISION_DENY:
		return core.PermissionStateDeny, nil
	case appv1.PermissionDecision_PERMISSION_DECISION_NONE:
		return core.PermissionStateNone, nil
	default:
		return "", invalidArgument("decision is required")
	}
}

func corePermissionTargetScope(scope *appv1.PermissionScope) core.PermissionTargetScope {
	if scope == nil {
		return core.PermissionTargetScope{}
	}
	switch scope.GetKind() {
	case appv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_GROUP:
		return core.PermissionTargetScope{Kind: core.MatrixScopeGroup, ID: scope.GetId()}
	case appv1.PermissionScopeKind_PERMISSION_SCOPE_KIND_ROOM:
		return core.PermissionTargetScope{Kind: core.MatrixScopeRoom, ID: scope.GetId()}
	default:
		return core.PermissionTargetScope{}
	}
}

File: apps/frontend/src/lib/api-client/permissions.ts
import { authHeaders, createChattoClient } from './connect.js';
import { AdminPermissionService } from '@chatto/api-types/admin/v1/permissions_connect';
import {
  PermissionDecision,
  PermissionScopeKind,
  type PermissionMatrixCell as APIPermissionMatrixCell,
  type PermissionMatrixScope as APIPermissionMatrixScope,
  type PermissionDecisionUpdate as APIPermissionDecisionUpdate,
  type RolePermissionMatrix as APIRolePermissionMatrix,
  type ScopedPermissionDecision as APIScopedPermissionDecision,
  type TierRole as APITierRole,
  type TierRoles as APITierRoles,
  type UserPermissionMatrix as APIUserPermissionMatrix
} from '@chatto/api-types/admin/v1/permissions_pb';

export type PermissionAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type PermissionState = 'allow' | 'deny' | 'neutral';
export type MatrixDecision = 'ALLOW' | 'DENY' | 'NONE';
export type MatrixScopeKind = 'SERVER' | 'GROUP' | 'ROOM' | 'DM';

export type PermissionScope =
  | { tier: 'server' }
  | { tier: 'dm' }
  | { tier: 'group'; groupId: string }
  | { tier: 'room'; roomId: string };

export type TierPermissions = {
  permissions: string[];
  permissionDenials: string[];
};

export type TierRole = {
  roleName: string;
  displayName: string;
  description: string;
  isSystem: boolean;
  position: number;
  pingable: boolean;
  override: TierPermissions;
  inheritedAllows: string[];
  inheritedDenials: string[];
};

export type TierRoles = {
  applicablePermissions: string[];
  roles: TierRole[];
};

export type MatrixScope = {
  id: string;
  label: string;
  kind: MatrixScopeKind;
  parentGroupId: string;
};

export type MatrixCell = {
  permission: string;
  scopeId: string;
  override: MatrixDecision;
  effective: MatrixDecision;
  allowPermitted?: boolean;
};

export type MatrixData = {
  applicablePermissions: string[];
  scopes: MatrixScope[];
  cells: MatrixCell[];
};

/** Scope counts, rather than permission-cell counts. */
export type PermissionScopePage = { totalCount: number; hasMore: boolean };
export type PermissionReadOptions = {
  signal?: AbortSignal;
  page?: { limit?: number; offset?: number };
  scope?: PermissionScope;
};

export type RolePermissionMatrix = MatrixData & {
  roleName: string;
  page: PermissionScopePage;
};

export type UserPermissionMatrix = MatrixData & {
  userId: string;
  page: PermissionScopePage;
};

export type PermissionDecisionEntry = {
  permission: string;
  scope: PermissionScope;
  override: MatrixDecision;
  effective: MatrixDecision;
};

export type PermissionDecisionUpdate = {
  permission: string;
  scope: PermissionScope;
  decision: MatrixDecision;
};

export type RolePermissionDecisions = {
  roleName: string;
  decisions: PermissionDecisionEntry[];
  scopes: PermissionScope[];
  page: PermissionScopePage;
};

export type UserPermissionDecisions = {
  userId: string;
  decisions: PermissionDecisionEntry[];
  scopes: PermissionScope[];
  page: PermissionScopePage;
};

export function createPermissionAPI(config: PermissionAPIConfig) {
  const client = createChattoClient(AdminPermissionService, config);
  const headers = () => authHeaders(config);

  return {
    async getRolePermissionTierMatrix(
      input: {
        roomId?: string | null;
        groupId?: string | null;
      },
      options: { signal?: AbortSignal } = {}
    ): Promise<TierRoles | null> {
      const response = await client.getRolePermissionTierMatrix(
        {
          scope: apiTierMatrixScope(input)
        },
        { headers: headers(), ...(options.signal ? { signal: options.signal } : {}) }
      );
      return response.matrix ? tierRoles(response.matrix) : null;
    },

    async getRolePermissionMatrix(
      roleName: string,
      options: PermissionReadOptions = {}
    ): Promise<RolePermissionMatrix | null> {
      const response = await client.getRolePermissionMatrix(
        {
          roleName,
          includeDirectMessageScope: true,
          page: options.page,
          scope: options.scope ? apiScope(options.scope) : undefined
        },
        { headers: headers(), ...(options.signal ? { signal: options.signal } : {}) }
      );
      return response.matrix
        ? { ...rolePermissionMatrix(response.matrix), page: scopePage(response.page) }
        : null;
    },

    async listRolePermissionDecisions(
      roleName: string,
      options: PermissionReadOptions = {}
    ): Promise<RolePermissionDecisions> {
      const response = await client.listRolePermissionDecisions(
        {
          roleName,
          includeDirectMessageScope: true,
          page: options.page,
          scope: options.scope ? apiScope(options.scope) : undefined
        },
        { headers: headers(), signal: options.signal }
      );
      return {
        roleName: response.roleName,
        decisions: response.decisions.map(permissionDecisionEntry),
        scopes: response.scopes.map(permissionScope),
        page: scopePage(response.page)
      };
    },

    async getUserPermissionMatrix(
      userId: string,
      options: PermissionReadOptions = {}
    ): Promise<UserPermissionMatrix | null> {
      const response = await client.getUserPermissionMatrix(
        {
          userId,
          includeDirectMessageScope: true,
          page: options.page,
          scope: options.scope ? apiScope(options.scope) : undefined
        },
        { headers: headers(), ...(options.signal ? { signal: options.signal } : {}) }
      );
      return response.matrix
        ? { ...userPermissionMatrix(response.matrix), page: scopePage(response.page) }
        : null;
    },

    async listUserPermissionDecisions(
      userId: string,
      options: PermissionReadOptions = {}
    ): Promise<UserPermissionDecisions> {
      const response = await client.listUserPermissionDecisions(
        {
          userId,
          includeDirectMessageScope: true,
          page: options.page,
          scope: options.scope ? apiScope(options.scope) : undefined
        },
        { headers: headers(), signal: options.signal }
      );
      return {
        userId: response.userId,
        decisions: response.decisions.map(permissionDecisionEntry),
        scopes: response.scopes.map(permissionScope),
        page: scopePage(response.page)
      };
    },

    async setRolePermission(input: {
      roleName: string;
      scope: PermissionScope;
      permission: string;
      state: PermissionState;
    }): Promise<PermissionDecisionUpdate> {
      const response = await client.setRolePermission(
        {
          roleName: input.roleName,
          permission: input.permission,
          decision: apiDecision(input.state),
          scope: apiScope(input.scope)
        },
        { headers: headers() }
      );
      return permissionDecisionUpdate(response.decision);
    },

    async setUserPermission(input: {
      userId: string;
      scope: PermissionScope;
      permission: string;
      state: PermissionState;
    }): Promise<PermissionDecisionUpdate> {
      const response = await client.setUserPermission(
        {
          userId: input.userId,
          permission: input.permission,
          decision: apiDecision(input.state),
          scope: apiScope(input.scope)
        },
        { headers: headers() }
      );
      return permissionDecisionUpdate(response.decision);
    }
  };
}

export type PermissionAPI = ReturnType<typeof createPermissionAPI>;

function tierRoles(matrix: APITierRoles): TierRoles {
  return {
    applicablePermissions: [...matrix.applicablePermissions],
    roles: matrix.roles.map(tierRole)
  };
}

function tierRole(role: APITierRole): TierRole {
  const apiRole = role.role;
  if (!apiRole) {
    throw new Error('permission tier role response did not include role metadata');
  }
  return {
    roleName: apiRole.name,
    displayName: apiRole.displayName,
    description: apiRole.description,
    isSystem: apiRole.isSystem,
    position: apiRole.position,
    pingable: apiRole.pingable,
    override: {
      permissions: [...(role.override?.permissions ?? [])],
      permissionDenials: [...(role.override?.permissionDenials ?? [])]
    },
    inheritedAllows: [...role.inheritedAllows],
    inheritedDenials: [...role.inheritedDenials]
  };
}

function rolePermissionMatrix(matrix: APIRolePermissionMatrix): Omit<RolePermissionMatrix, 'page'> {
  return {
    roleName: matrix.roleName,
    applicablePermissions: [...matrix.applicablePermissions],
    scopes: matrix.scopes.map(matrixScope),
    cells: matrix.cells.map(matrixCell)
  };
}

function userPermissionMatrix(matrix: APIUserPermissionMatrix): Omit<UserPermissionMatrix, 'page'> {
  return {
    userId: matrix.userId,
    applicablePermissions: [...matrix.applicablePermissions],
    scopes: matrix.scopes.map(matrixScope),
    cells: matrix.cells.map(matrixCell)
  };
}

function matrixScope(scope: APIPermissionMatrixScope): MatrixScope {
  return {
    id: scope.id,
    label: scope.label,
    kind: scopeKind(scope.kind),
    parentGroupId: scope.parentGroupId
  };
}

function matrixCell(cell: APIPermissionMatrixCell): MatrixCell {
  return {
    permission: cell.permission,
    scopeId: cell.scopeId,
    override: matrixDecision(cell.override),
    effective: matrixDecision(cell.effective),
    ...(cell.allowPermitted !== undefined ? { allowPermitted: cell.allowPermitted } : {})
  };
}

function permissionDecisionEntry(decision: APIScopedPermissionDecision): PermissionDecisionEntry {
  return {
    permission: decision.permission,
    scope: permissionScope(decision.scope),
    override: matrixDecision(decision.override),
    effective: matrixDecision(decision.effective)
  };
}

function permissionDecisionUpdate(
  decision: APIPermissionDecisionUpdate | undefined
): PermissionDecisionUpdate {
  if (!decision) {
    throw new Error('permission write response did not include a decision');
  }
  return {
    permission: decision.permission,
    scope: permissionScope(decision.scope),
    decision: matrixDecision(decision.decision)
  };
}

function permissionScope(
  scope: { kind: PermissionScopeKind; id: string } | undefined
): PermissionScope {
  if (scope?.kind === PermissionScopeKind.GROUP) {
    return { tier: 'group', groupId: scope.id };
  }
  if (scope?.kind === PermissionScopeKind.ROOM) {
    return { tier: 'room', roomId: scope.id };
  }
  if (scope?.kind === PermissionScopeKind.DM) return { tier: 'dm' };
  if (scope?.kind === PermissionScopeKind.SERVER) return { tier: 'server' };
  throw new Error(`unsupported permission scope kind: ${scope?.kind ?? 'missing'}`);
}

function scopeKind(kind: PermissionScopeKind): MatrixScopeKind {
  if (kind === PermissionScopeKind.GROUP) return 'GROUP';
  if (kind === PermissionScopeKind.ROOM) return 'ROOM';
  if (kind === PermissionScopeKind.DM) return 'DM';
  if (kind === PermissionScopeKind.SERVER) return 'SERVER';
  throw new Error(`unsupported permission matrix scope kind: ${kind}`);
}

function matrixDecision(decision: PermissionDecision): MatrixDecision {
  if (decision === PermissionDecision.ALLOW) return 'ALLOW';
  if (decision === PermissionDecision.DENY) return 'DENY';
  return 'NONE';
}

function apiDecision(state: PermissionState): PermissionDecision {
  if (state === 'allow') return PermissionDecision.ALLOW;
  if (state === 'deny') return PermissionDecision.DENY;
  return PermissionDecision.NONE;
}

function apiScope(scope: PermissionScope): {
  kind: PermissionScopeKind;
  id: string;
} {
  if (scope.tier === 'group') {
    return { kind: PermissionScopeKind.GROUP, id: scope.groupId };
  }
  if (scope.tier === 'room') {
    return { kind: PermissionScopeKind.ROOM, id: scope.roomId };
  }
  if (scope.tier === 'dm') {
    return { kind: PermissionScopeKind.DM, id: '' };
  }
  return { kind: PermissionScopeKind.SERVER, id: '' };
}

function apiTierMatrixScope(input: { roomId?: string | null; groupId?: string | null }): {
  kind: PermissionScopeKind;
  id: string;
} {
  if (input.roomId) {
    return { kind: PermissionScopeKind.ROOM, id: input.roomId };
  }
  if (input.groupId) {
    return { kind: PermissionScopeKind.GROUP, id: input.groupId };
  }
  return { kind: PermissionScopeKind.SERVER, id: '' };
}

function scopePage(
  page: { totalCount: bigint; hasMore: boolean } | undefined
): PermissionScopePage {
  if (!page) throw new Error('permission response did not include scope page metadata');
  return { totalCount: Number(page.totalCount), hasMore: page.hasMore };
}
